[SVG] Interval should not be created if a value in begin-value-list doesn't have a matching value in end-value-list

Ahmad-S792 opened the pull request:
#### 0d691ef763d7d68ec4d4e3d795145b8b20325343
<pre>
[SVG] Interval should not be created if a value in begin-value-list doesn&apos;t have a matching value in end-value-list

<a href="https://bugs.webkit.org/show_bug.cgi?id=257105">https://bugs.webkit.org/show_bug.cgi?id=257105</a>

Reviewed by Simon Fraser.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: <a href="https://src.chromium.org/viewvc/blink?view=revision&amp">https://src.chromium.org/viewvc/blink?view=revision&amp</a>;revision=157739

Here <a href="http://www.w3.org/TR/SMIL3/smil-timing.html#q90">http://www.w3.org/TR/SMIL3/smil-timing.html#q90</a> , it says
that if all ends are before the current begin, then it is a bad interval
and current begin will be discarded.
Existing code is treating it as a valid interval with end as indefinite.
Modified the logic to be in line with the pseudo code.

* Source/WebCore/svg/animation/SVGSMILElement.cpp:
(SVGSMILElement::findInstanceTime): change &apos;return&apos; value
(SVGSMILElement::resolveNextInterval): Change to &apos;bool&apos;
(SVGSMILElement::seekToIntervalCorrespondingToTime): Modify &apos;resolveNextInterval&apos; in &apos;if&apos; condition
* Source/WebCore/svg/animation/SVGSMILElement.h: &apos;resolveNextInterval&apos; to bool
* LayoutTests/resources/SVGAnimationTestCase-testharness.js: Move this to &apos;resource&apos; folder as well
* LayoutTests/svg/animations/animate-no-matching-end-value.html: Add Test Case
* LayoutTests/svg/animations/animate-no-matching-end-value-expected.txt: Add Test Case Expectation

Canonical link: <a href="https://commits.webkit.org/264635@main">https://commits.webkit.org/264635@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7ea222610612a723990ff303a9edd613ca6b8a8e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7360 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7617 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7791 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8987 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7565 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7370 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/9345 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7542 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10456 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7489 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/8461 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6803 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9096 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/5785 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6698 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14423 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7148 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6807 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/9988 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7296 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5960 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6649 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/6609 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1976 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10854 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7033 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/svg/animation/SVGSMILElement.h:
```diff
@@ -137,7 +137,7 @@ class SVGSMILElement : public SVGElement {
     SMILTime findInstanceTime(BeginOrEnd, SMILTime minimumTime, bool equalsMinimumOK) const;
     void resolveFirstInterval();
-    void resolveNextInterval();
+    bool resolveNextInterval();
```
I'd like to see a comment saying what the return value means (or make it an enum).

Ahmad-S792 replied:
I have few follow-up blink merges planned, so I think those might change this entirely. :-)

Ahmad-S792 replied:
Since I have future plans to merge few more Blink patches in SMIL to make WebKit better. I am going to take it as 'RESOLVED' for time being.

smfr commented on LayoutTests/resources/SVGAnimationTestCase-testharness.js:
```diff
@@ -0,0 +1,102 @@
+// NOTE(edvardt):
```
This kind of comment isn't useful. Don't know who `edvardt` is, and a reference to "the old SVGAnimationTestCase.js" isn't useful to anyone reading this in the future.

Ahmad-S792 replied:
It is just a copy taken from Chromium so I kept it as is. If you want, I am happy to remove.

Ahmad-S792 replied:
@smfr - It is present in WPT script as well. [Link](https://github.com/web-platform-tests/wpt/blob/35283827b1658ca4856238bb798d53ece16f1445/resources/SVGAnimationTestCase-testharness.js#L4)

Ahmad-S792 committed:
[SVG] Interval should not be created if a value in begin-value-list doesn't have a matching value in end-value-list

https://bugs.webkit.org/show_bug.cgi?id=257105

Reviewed by Simon Fraser.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: https://src.chromium.org/viewvc/blink?view=revision&revision=157739

Here http://www.w3.org/TR/SMIL3/smil-timing.html#q90 , it says
that if all ends are before the current begin, then it is a bad interval
and current begin will be discarded.
Existing code is treating it as a valid interval with end as indefinite.
Modified the logic to be in line with the pseudo code.

* Source/WebCore/svg/animation/SVGSMILElement.cpp:
(SVGSMILElement::findInstanceTime): change 'return' value
(SVGSMILElement::resolveNextInterval): Change to 'bool'
(SVGSMILElement::seekToIntervalCorrespondingToTime): Modify 'resolveNextInterval' in 'if' condition
* Source/WebCore/svg/animation/SVGSMILElement.h: 'resolveNextInterval' to bool
* LayoutTests/resources/SVGAnimationTestCase-testharness.js: Move this to 'resource' folder as well
* LayoutTests/svg/animations/animate-no-matching-end-value.html: Add Test Case
* LayoutTests/svg/animations/animate-no-matching-end-value-expected.txt: Add Test Case Expectation

Canonical link: https://commits.webkit.org/264635@main
